# Lab 00 - Pipeline check

**下午第一節，約 40 分鐘。四段路徑走一遍。**

Dashboard：<http://localhost:3000/d/aiops-lab00>（標題 `Lab 00 - Pipeline check`）

這堂課沒有異常偵測。要證明的是一件事：你在 notebook 裡算出來的一個數字，能不能出現在 Grafana 的 panel 上。這聽起來像雜事，但下午其餘的每一個 lab 都站在這條路徑上，路徑不通的話，Lab 01 的 baseline 和 Lab 02 的 alert rule 全部只是本機的一堆 float。

課程結束時你要能做到兩件事。第一，看著 panel 上的線跟著你改的參數動。第二，當它沒有動的時候，指出四段路徑裡壞的是哪一段，而且是從 dashboard 上看出來的，不是靠別人告訴你。

這個 notebook 裡不畫圖。今天要看的圖全部在 Grafana 上。

## 這條路徑長什麼樣子

```text
hop 1   notebook            ->  outputs/prometheus-dropzone/current_results.csv
                                （同時寫出 current_results.manifest.json）
hop 2   results exporter    ->  infra/python_results_exporter.py 在 :8010/metrics 暴露
hop 3   Prometheus          ->  scrape job "python-results-exporter"，每 5 秒抓一次
hop 4   Grafana             ->  查詢 aiops_python_result{column="...", port_id="..."}
```

這四段之後會一直用 hop 1 到 hop 4 稱呼，請記住編號。這個 notebook 的 Python 停在 hop 1，其餘三段的操作在 Grafana 的介面上做，步驟就寫在下面的 markdown cell 裡。

直接讓 Grafana 讀 CSV 會快很多，但那不是任何 production 系統的長相。真實環境裡沒有人會把 dashboard 指到一個檔案上，因為檔案沒有 label、沒有保存期限、沒有 scrape 失敗的紀錄，也沒有辦法讓 alert rule 對它做評估。這個 workshop 把 production 的路徑原封不動保留下來，只是把每一段縮短：drop zone 取代了訊息佇列，單一 exporter 取代了一整排 agent。形狀一樣，所以你在這裡學到的 debug 順序，回到公司的監控系統仍然適用。

manifest 的內容與它對 exporter 的約束，第 4 節「送上路」會展開。

## 1. 啟動四個服務

每一個都要佔一個終端機視窗，因為其中三個是前景程序，關掉視窗就等於停掉服務。先確認你在 repository 根目錄，而且 conda 環境已經啟動。

```bash
conda activate aiops-anomaly-zero-to-hero
cd <repo root>
```

終端機 1，node_exporter，提供你這台機器的真實網路指標：

```bash
node_exporter
```

終端機 2，Prometheus。這一行是本節的重點，等一下會解釋為什麼：

```bash
prometheus --config.file=infra/prometheus/prometheus.macos.yml
```

Linux 換成 `prometheus.linux.yml`，Windows 換成 `prometheus.windows.yml`，三個檔案的 scrape target 完全一樣。

終端機 3，results exporter。`REPLAY_SPEED_X` 決定模擬時鐘走多快，720 表示每一個真實秒推進 720 個模擬秒：

```bash
REPLAY_SPEED_X=720 python infra/python_results_exporter.py
```

終端機 4，Grafana。macOS 用 `brew services start grafana`，Linux 用 `sudo systemctl start grafana-server`。Grafana 是背景服務，這個視窗之後可以拿來做別的事。

### 那個會讓你浪費半小時的坑

macOS 上最自然的做法是 `brew services start prometheus`。不要這樣做。Homebrew 的 service 定義會去讀 `/opt/homebrew/etc/prometheus.yml`，那個檔案是 Homebrew 自己的預設值，裡面只有 Prometheus 監控自己的那個 job，從來沒聽過 `python-results-exporter`。

這個失敗模式的惡毒之處在於它看起來完全正常。Prometheus 起來了，網頁打得開，`up` 查得到東西，exporter 的 `:8010/metrics` 也有輸出，Grafana 也活著。四個服務各自檢查都是好的，但 hop 3 根本沒有把 hop 2 接上 hop 4，dashboard 就一直是空的。

`wk.check_stack()` 專門認得這一種情況。它除了看服務活著沒有，還會去比對 Prometheus 執行中的 config 裡有沒有那個 job。下一格的 `wk.bootstrap()` 內部就會跑 `check_stack`，執行它，然後讀印出來的東西。

In [1]:
import json
import sys
import pathlib

PROJECT_ROOT = pathlib.Path.cwd()
while PROJECT_ROOT != PROJECT_ROOT.parent and not (PROJECT_ROOT / "environment.yml").exists():
    PROJECT_ROOT = PROJECT_ROOT.parent
sys.path.insert(0, str(PROJECT_ROOT / "infra"))

import numpy as np
import pandas as pd

import aiopskit as wk

print(f"pandas        {pd.__version__}")
print(f"numpy         {np.__version__}")
print()

project_root = wk.bootstrap()

pandas        3.0.3
numpy         2.4.6

project root  /Users/kevin/Library/CloudStorage/GoogleDrive-kevin880987@gmail.com/我的雲端硬碟/workshop/workbench/project/202608-aia-network-monitoring-aiops/aiops-anomaly-zero-to-hero
aiopskit      1.0.0

component             status    detail
------------------------------------------------------------------------------
Prometheus            up        http://localhost:9090
results exporter      up        http://localhost:8010/metrics
Grafana               up        http://localhost:3000  datasource=prometheus
  scrape prometheus     up        
  scrape rrd-exporter   down      
  scrape node-exporter  up        
  aiops_python_result             0 series in Prometheus

Prometheus is running but has no 'python-results-exporter' scrape job, so nothing you
  publish can reach Grafana. This happens when Prometheus was started from a
  package default config instead of the repository one. Stop it and restart with:
    prometheus --config.file=infra/prome

### 讀它印出來的東西

一台踩到坑的機器印出來的是這樣：

```text
component             status    detail
------------------------------------------------------------------------------
Prometheus            up        http://localhost:9090
results exporter      up        http://localhost:8010/metrics
Grafana               up        http://localhost:3000  datasource=prometheus
  scrape prometheus     up
  scrape rrd-exporter   down
  scrape node-exporter  up
  aiops_python_result             0 series in Prometheus

Prometheus is running but has no 'python-results-exporter' scrape job, so nothing you
  publish can reach Grafana. This happens when Prometheus was started from a
  package default config instead of the repository one. Stop it and restart with:
    prometheus --config.file=infra/prometheus/prometheus.macos.yml
  (use prometheus.linux.yml or prometheus.windows.yml as appropriate)
```

三個服務全部 up，`aiops_python_result` 卻是 0 series。看到這段訊息就回終端機 2，Ctrl-C 停掉 Prometheus，用 `--config.file` 重新啟動。

`scrape rrd-exporter down` 不用理會。那是自學路線用的 exporter，工作坊路線不需要它，Prometheus 對抓不到的 target 只會安靜記一筆。

修好之後再跑一次上面那一格，最後幾行應該變成 `all services up`。**在這一行出現之前不要往下走。** 後面每一節都預設這四段是通的，帶著壞掉的 hop 3 繼續，你只會在第 4 節得到一個空白的 panel，然後花二十分鐘懷疑自己的 Python。

## 2. 用一行指令把整個工作坊裝進 Grafana

手動在 UI 裡建 datasource、建 folder、貼三份 dashboard JSON、設三條 alert rule，是二十分鐘的機械勞動，而且十個人會有十種手滑的方式。這個 repository 把那些都寫成程式：

```bash
python infra/setup_grafana.py
```

它建立五類物件。uid `prometheus` 的 Prometheus datasource 底下掛一個名為 **AIOps Workshop** 的 folder（uid `aiops-workshop`），folder 裡放三份 dashboard `aiops-lab00`、`aiops-lab01`、`aiops-lab02`，以及三條全部歸在 `lab02` group 的 alert rule。最後是 mute timing `aiops-maintenance-window`，每天 00:30 到 05:00。

這個腳本是 idempotent 的，重跑不會製造重複物件。真的把某個東西改壞了，`python infra/setup_grafana.py --reset` 會先刪掉工作坊的物件再重建。`--dry-run` 只把 dashboard JSON 寫到 `infra/grafana/dashboards/`，完全不碰 Grafana。

有一個設計決定現在要先講清楚，因為 Lab 02 會用到。腳本推送 alert rule 的時候刻意送出 `X-Disable-Provenance: true`，效果是 Grafana 把這三條規則當成手工建立的，而不是被 provisioning 檔案管理的。**被 provisioning 管理的規則在 UI 上是唯讀的，這三條不是。** 你可以直接在 Alerting 介面裡改門檻、改 `for` 的長度、改 severity，改完立刻生效。Lab 02 的後半段就是在做這件事：在 UI 上改掉那條規則的門檻和 pending period，看狀態機怎麼從 Normal 走到 Pending 再走到 Firing，然後把值改回去。腳本的角色是把一個能動的範例放在你面前，實際的學習發生在你動手改它的時候。

指令跑完會印出四個連結。現在只需要第一個。

## 3. 讀懂四個 health tile

打開 `http://localhost:3000/d/aiops-lab00`，或者從左側導覽列 **Dashboards** 進去，在 **AIOps Workshop** folder 裡找 **Lab 00 - Pipeline check**。

最上面是一個叫 **What you are looking at** 的 text panel，寫的是這個 dashboard 要回答的問題。底下一整排四個 stat tile，是這一節的主角。它們現在多半是空的或是 0，那是正常的，因為你還沒有 publish 任何東西。先弄懂每一個 tile 在證明什麼，等一下數字跳出來的時候才知道自己在看什麼。

**Exporter up** 查的是 `up{job="python-results-exporter"}`。這個 metric 有一個關鍵性質：它由 **Prometheus 自己在每次 scrape 之後寫下來**，exporter 沒有參與。抓到就記 1，抓不到就記 0。所以這個 tile 證明的是 hop 3 的健康，而且它是唯一一個在 exporter 死掉之後還會繼續有值的 tile。

**Rows loaded** 查 `aiops_result_rows`，由 exporter 在讀完 CSV 之後設定。它證明 hop 1 和 hop 2 都活著：drop zone 裡有一個檔案，而且 exporter 成功把它 parse 成 DataFrame。發佈 toy signal 之後這裡會是 1440。跑完 Lab 02 之後會是 43200，也就是整份課程資料集的列數。

**Replay speed** 這一塊直接映射你在終端機 3 設的那個環境變數，metric 名字是 `aiops_replay_speed_x`，應該是 720。如果看到 3600，表示 `REPLAY_SPEED_X=` 沒有帶到，exporter 用了預設值，dashboard 上的資料會跑得比講師的快五倍。

最後一塊 **Replay progress** 走 `aiops_replay_progress`。metric 本身是 0 到 1，但 tile 的 unit 設成 `percentunit`，所以螢幕上顯示的是 `45.3%` 這樣的百分比，不是 `0.453`；走到 100% 之後 wrap 回 0 重播。它證明 exporter 的主迴圈還在轉，不是一個卡住但還佔著 port 的殭屍程序。四個 tile 裡只有這一個會自己動，盯著看幾秒就知道。

replay 的理由在資料集的長度。課程資料集橫跨 2026-02-01 到 2026-03-02，整整 30 天，cadence 300 秒。如果照原本的時間軸畫，你需要在螢幕前坐一個月才看得到第二個事件。720 倍之下，30 天的資料走完一輪需要 60 個真實分鐘，剛好是一堂課的長度。代價是資料的 wall-clock 時戳永遠是「現在」，所以**這些 dashboard 的時間範圍一律是相對的**，Lab 00 開在 `now-15m`，Lab 01 和 Lab 02 開在 `now-30m`。把時間選擇器改成某個絕對日期，你會得到一張空圖，而那不是 pipeline 壞了。

## 4. 造一條刻意做得很笨的訊號，然後送上路

`wk.synthetic_wave()` 給一條正弦波、一點高斯雜訊，再在大約 70% 的位置注入一段 +45 的階躍。笨是重點：等一下 panel 上出現東西的時候，你要能一眼認出那就是這裡算的那條，而不是某個看起來也很像時間序列的東西。

三個參數各有理由。`CADENCE_S = 300` 與課程資料集一致，一天 288 個樣本，不是 1440 個。`N_SAMPLES = 1440` 讓模擬時間橫跨 5 天，在 `REPLAY_SPEED_X=720` 之下大約 10 個真實分鐘走完一輪，所以 dashboard 的 15 分鐘視窗裡永遠看得到至少一次完整的波形。最後那個 24 小時的 rolling median 由 `BASELINE_WINDOW = 288` 決定；注入的階躍只有 116 個樣本，佔不到視窗的一半，中位數因此不會被它拉走。Lab 01 會把這件事講完整，現在先看它發生。

下一格建完訊號之後先印一份 profile。這堂 lab 沒有真實遙測資料，profile 因此很短，但習慣要從這裡開始養：任何一份要進 pipeline 的表，先看形狀、cadence、時間跨度、每一個要發佈的欄位的 null 數與數值範圍，再談要對它做什麼。Lab 01 拿到 43,200 列真實資料時，第一格做的是同一件事，只是欄位多得多。

In [2]:
N_SAMPLES = 1440        # 1440 x 300 s = 5 天模擬時間
CADENCE_S = 300         # 與 data/synthetic/synthetic_rrd_metrics.csv 相同
BASELINE_WINDOW = 288   # 288 個樣本 = 24 小時
PUBLISHED = ["toy_value", "toy_baseline"]

toy = wk.synthetic_wave(n=N_SAMPLES, cadence_s=CADENCE_S)
toy["toy_baseline"] = (
    toy["toy_value"].rolling(BASELINE_WINDOW, min_periods=24).median().bfill()
)

# ---- profile: shape, cadence, span, range, before anything is published -------
cadence = wk.infer_cadence(toy["timestamp"])
span_s = (toy["timestamp"].max() - toy["timestamp"].min()).total_seconds()

print(f"shape             {toy.shape[0]:,} rows x {toy.shape[1]} columns")
print(f"columns           {', '.join(toy.columns)}")
print(f"span              {toy['timestamp'].min()}  ->  {toy['timestamp'].max()}")
print(f"cadence           {cadence:.0f} s per sample ({86400 / cadence:.0f} samples/day)")
print(f"sim span          {span_s / 86400:.2f} days")
print(f"one replay pass   {span_s / 720 / 60:.2f} real minutes at REPLAY_SPEED_X=720")
print(f"label inventory   device_id={toy['device_id'].unique().tolist()} "
      f"port_id={toy['port_id'].unique().tolist()} "
      f"port_role={toy['port_role'].unique().tolist()}")
print(f"event labels      {toy['event_label'].value_counts().to_dict()}")
print(f"above 85          {int((toy['toy_value'] > 85).sum())} of {len(toy)} samples "
      f"({(toy['toy_value'] > 85).mean():.1%})")
print()

profile = pd.DataFrame({
    "nulls": toy[PUBLISHED].isna().sum(),
    "zero_frac": (toy[PUBLISHED] == 0).mean().round(3),
    "min": toy[PUBLISHED].min().round(1),
    "median": toy[PUBLISHED].median().round(1),
    "p95": toy[PUBLISHED].quantile(0.95).round(1),
    "max": toy[PUBLISHED].max().round(1),
})
print(profile.to_string())
print()
print(toy.head(3).to_string(index=False))

shape             1,440 rows x 7 columns
columns           timestamp, device_id, port_id, port_role, event_label, toy_value, toy_baseline
span              2026-07-18 10:20:00  ->  2026-07-23 10:15:00
cadence           300 s per sample (288 samples/day)
sim span          5.00 days
one replay pass   9.99 real minutes at REPLAY_SPEED_X=720
label inventory   device_id=['toy-device'] port_id=['toy-port'] port_role=['demo']
event labels      {'normal': 1440}
above 85          116 of 1440 samples (8.1%)

              nulls  zero_frac   min  median   p95    max
toy_value         0        0.0  25.2    52.1  94.7  107.4
toy_baseline      0        0.0  34.0    58.0  70.2   71.0

          timestamp  device_id  port_id port_role event_label  toy_value  toy_baseline
2026-07-18 10:20:00 toy-device toy-port      demo      normal  50.002460     50.748146
2026-07-18 10:25:00 toy-device toy-port      demo      normal  50.772143     50.748146
2026-07-18 10:30:00 toy-device toy-port      demo      norma

### 讀這份 profile

兩欄都沒有 null，因為 `bfill()` 把 rolling median 開頭那段還沒滿窗的樣本補掉了；zero fraction 是 0，所以沒有「用 0 冒充缺值」的問題。cadence 是從時戳量出來的 300 秒，一天 288 個樣本，`BASELINE_WINDOW = 288` 這個選擇因此等於 24 小時，同一個換算 Lab 01 的每一個 rolling window 都會再用一次。`toy_value` 的 max 超過 100，`toy_baseline` 的 max 停在 71.0，兩者中間隔著 dashboard 上那條畫在 85 的 threshold line，所以待會兒圖上只有 `toy_value` 會衝出紅線。這就是 profile 的用途：先知道數值落在哪裡，才知道門檻該畫在哪裡。

### 送上路

`wk.publish()` 寫兩個檔案。`outputs/workshop/lab00_toy_signal.csv` 是這個 lab 的存檔，不會被別的 lab 覆蓋；`outputs/prometheus-dropzone/current_results.csv` 是 exporter 盯著的那一份。旁邊還會放一個 manifest，明確宣告哪幾欄是數值、哪幾欄當 label，exporter 因此不必用猜的。下一格會把 manifest 整份印出來，那就是 hop 1 交給 hop 2 的合約。

`labels=["device_id", "port_id"]` 決定哪些欄位會變成 Prometheus label。沒有列出來的 `port_role` 和 `event_label` 會以空字串輸出。這是 prometheus_client 的規則：Gauge 的 label 集合在建立時就固定了，之後不能少一個，只能填空。

drop zone 一次只放得下一個 lab 的結果。等一下跑 Lab 01 的 publish，這條 toy signal 就會從 Prometheus 上消失，這是這種設計的直接後果，不是故障。Lab 02 為此做了特別處理，它會把 Lab 01 和 Lab 02 的欄位聯集一起發佈，讓兩張 dashboard 同時有資料。

In [3]:
manifest = wk.publish(
    toy,
    columns=PUBLISHED,
    name="lab00_toy_signal",
    labels=["device_id", "port_id"],
)

print()
print("manifest written next to the drop-zone CSV:")
print(json.dumps(manifest, indent=2))

published 1,440 rows as 'lab00_toy_signal'
  archive   outputs/workshop/lab00_toy_signal.csv
  drop zone outputs/prometheus-dropzone/current_results.csv
  columns   toy_value, toy_baseline
  query     aiops_python_result{column="toy_value"}

manifest written next to the drop-zone CSV:
{
  "source": "lab00_toy_signal",
  "value_columns": [
    "toy_value",
    "toy_baseline"
  ],
  "label_columns": [
    "device_id",
    "port_id"
  ],
  "rows": 1440,
  "sim_start": "2026-07-18 10:20:00",
  "sim_end": "2026-07-23 10:15:00",
  "generated_at": "2026-07-23T10:15:32+00:00"
}


### 回終端機 3，再回 Grafana

exporter 每 5 秒檢查一次檔案的 mtime，發現變了就重新載入，你會看到它印出：

```text
Reloaded .../current_results.csv (1,440 rows)
Value columns: toy_value, toy_baseline
Active labels: device_id, port_id
```

`Active labels` 那一行就是上一格 `labels=[...]` 的回聲。

現在回 Grafana。**Rows loaded** 應該跳成 1440，**Exporter up** 是 1，**Replay progress** 開始爬。往下捲到 **Toy signal published from the notebook**，最多再等十幾秒，一條線會從右邊長出來。

那條線上有一條紅色的水平虛線，在 85 的位置。那是 panel 上設的 threshold line，不是 alert，沒有人會因為它被叫起來。它只是把「85 以上算高」這個判斷畫在圖上。toy signal 的階躍區間量到的值介於 87.1 和 107.4 之間，所以每一輪 replay 都會有一小段線衝出這條紅線之上。

還有一條顏色不同的線是 `toy_baseline`，就是上面那個 288 個樣本的 rolling median。它的範圍是 34.0 到 71.0，永遠不會碰到 85。原因很具體：階躍持續 116 個樣本，佔不到 288 的一半，而中位數要被污染超過一半才會移動。這正是 Lab 01 要講的 robust baseline，你現在已經先看到它動作了一次。

一輪 toy signal 走完需要 10 個真實分鐘。等於說，開著不管的話，panel 上每十分鐘會出現一次那段階躍。

## 5. 故意把它弄壞

這是整堂 lab 真正的內容。你已經看過通的時候長什麼樣子，現在要看斷的時候長什麼樣子，而且要在有人告訴你答案之前先自己讀出來。

切到終端機 3，按 Ctrl-C。exporter 會印出 `Stopped.` 然後結束。

回到 Grafana，等 10 到 15 秒讓 dashboard 刷新，然後仔細看那四個 tile。你會看到一個很特別的組合：

**Exporter up 變成 0，另外三個 tile 變成 No data。**

停一下，想清楚為什麼會是這樣，這是今天最值得帶走的一個觀察。`aiops_result_rows`、`aiops_replay_speed_x`、`aiops_replay_progress` 這三個都是 exporter 自己產生的 metric，程序死了就沒有人再產生它們，Prometheus 在下一次 scrape 失敗時替它們寫入 staleness marker，instant query 從此查不到值，stat panel 顯示 No data。而 `up` 由 Prometheus 在 scrape 失敗的當下自己合成出來，所以它不但沒有消失，還準確地報告了 0。

一個報 0，三個報 No data，而報 0 的那個正好是唯一不由 exporter 產生的。這個不對稱可以直接讀成診斷結論：**hop 2 死了，hop 3 還活著**。

再往下看 **Toy signal published from the notebook**。線停在最後一個值不再往前，右側慢慢空出來。這是 series 變 stale 的樣子，跟「數值掉到 0」在圖上完全不同，也跟「panel 從頭到尾都是空的」不同。三種空法對應三種不同的故障，把它們分清楚，日後在自己的系統上省下的時間比這堂課還長。

現在重新啟動它：

```bash
REPLAY_SPEED_X=720 python infra/python_results_exporter.py
```

十幾秒之內，四個 tile 全部回來。注意 **Replay progress** 從 0 開始重新爬，因為 exporter 的 `wall_start` 在啟動時重設，replay 從資料集的第一列重播。同樣地，toy signal 的波形也從頭開始，那段階躍要再等大約七分鐘才會出現。

### 症狀對照表

把它記住，或者留在螢幕上。

| 你在 dashboard 上看到的 | 壞的是 | 怎麼確認 |
| --- | --- | --- |
| Exporter up = 0，其餘三個 tile No data | hop 2，exporter 程序沒在跑 | `curl -s localhost:8010/metrics \| head`，沒有回應就是它 |
| Exporter up 顯示 **No data**（不是 0） | hop 3，Prometheus config 裡沒有這個 job | `wk.check_stack()`，或開 `localhost:9090/targets` 找不找得到 `python-results-exporter` |
| 四個 tile 都正常，Rows loaded 是 0 或舊數字 | hop 1，drop zone 沒有新檔案 | `ls -l outputs/prometheus-dropzone/`，比對 mtime 和 manifest 的 `generated_at` |
| tile 都正常，但時間序列 panel 空白 | hop 4，query 或時間範圍不對 | 到 Explore 貼同一段 PromQL，查得到就是 panel 設定問題 |
| 線停住不再往前 | hop 2 剛剛斷，Prometheus 已寫入 staleness marker | 看 Exporter up 是 0 還是 1 |
| 圖整段空白，且時間選擇器是絕對日期 | 沒有壞 | replay 的資料時戳永遠在「現在」，改回 `now-15m` |

`up` 顯示 0 跟顯示 No data 的差別，是這張表裡最重要的一行。0 表示 Prometheus 知道要抓這個 target，抓了，失敗了。No data 表示 Prometheus 從來不知道有這個 target。前者一眼就看得出來，後者需要你注意到「這個 tile 為什麼沒有值」，而人的眼睛對空白遠不如對紅色數字敏感。第 1 節那個 Homebrew 的坑，症狀就是後者。

## 6. 改 notebook，看 panel 跟著動

前面證明的是路徑會傳送資料。這一步要證明的是路徑傳送的是**你的**資料，不是某個 demo 資料的快取。

下一格暴露兩個參數。`AMPLITUDE_X` 把整條曲線相對中心 50 的偏離量整體縮放，`EXTRA_STEP` 疊加在原本那段階躍上。改其中一個就好，動作要大到不需要瞇眼睛：`AMPLITUDE_X = 2.0`，或者 `EXTRA_STEP = 40.0`。`AMPLITUDE_X = 2.0` 之下超過 85 的樣本數會從 116 跳到 320。

改完之後**一定要重跑那一格**。它最後一段就是 publish，少跑的話 drop zone 還是舊檔案，panel 當然不動，然後你會開始懷疑 pipeline。

In [4]:
AMPLITUDE_X = 1.0    # 改這裡：1.0 是原始振幅，2.0 是兩倍
EXTRA_STEP = 0.0     # 改這裡：疊加在原本那段階躍上的額外高度
CENTER = 50.0

toy = wk.synthetic_wave(n=N_SAMPLES, cadence_s=CADENCE_S)
toy["toy_value"] = CENTER + (toy["toy_value"] - CENTER) * AMPLITUDE_X

step_index = toy.index[int(N_SAMPLES * 0.70):int(N_SAMPLES * 0.78)]
toy.loc[step_index, "toy_value"] += EXTRA_STEP

toy["toy_baseline"] = (
    toy["toy_value"].rolling(BASELINE_WINDOW, min_periods=24).median().bfill()
)

print(f"toy_value      {toy['toy_value'].min():6.1f} .. {toy['toy_value'].max():6.1f}")
print(f"toy_baseline   {toy['toy_baseline'].min():6.1f} .. {toy['toy_baseline'].max():6.1f}")
print(f"above 85       {int((toy['toy_value'] > 85).sum())} of {len(toy)} samples")
print()

manifest = wk.publish(
    toy,
    columns=PUBLISHED,
    name="lab00_toy_signal",
    labels=["device_id", "port_id"],
)

toy_value        25.2 ..  107.4
toy_baseline     34.0 ..   71.0
above 85       116 of 1440 samples

published 1,440 rows as 'lab00_toy_signal'
  archive   outputs/workshop/lab00_toy_signal.csv
  drop zone outputs/prometheus-dropzone/current_results.csv
  columns   toy_value, toy_baseline
  query     aiops_python_result{column="toy_value"}


盯著終端機 3，等 `Reloaded` 出現，再回 Grafana。panel 上的波形會在下一輪 replay 換掉。`AMPLITUDE_X = 2.0` 之下整條線的擺幅變成兩倍，谷底會壓到 0 附近，波峰則會有相當長的時間待在 85 那條紅線之上。

有一件事要一起看：`toy_baseline` 也跟著變了。它是從 `toy_value` 算出來的 rolling median，輸入變了輸出當然變。這件事本身很平凡，但它在 dashboard 上呈現的意思不平凡：**你改的是 Python，動的是牆上的圖**。Lab 01 之後要做的每一件事都是這個動作的放大版，只是把這條 rolling median 換成 Lab 01 那幾種彼此競爭的 baseline，把 toy signal 換成五個真實 port 的流量。

下一格印出把 PromQL 預先填好的 Grafana Explore 連結，第 7 節從那裡開始。

In [5]:
grafana = wk.GrafanaClient()
print("Grafana reachable:", grafana.healthy())
print()

for column in PUBLISHED:
    query = f'aiops_python_result{{column="{column}"}}'
    print(query)
    print(" ", grafana.explore_url(query, minutes=15))
    print()

print("Lab 00 dashboard   http://localhost:3000/d/aiops-lab00")

Grafana reachable: True

aiops_python_result{column="toy_value"}


  http://localhost:3000/explore?schemaVersion=1&orgId=1&panes=%7B%22aiops%22%3A+%7B%22datasource%22%3A+%22prometheus%22%2C+%22queries%22%3A+%5B%7B%22refId%22%3A+%22A%22%2C+%22expr%22%3A+%22aiops_python_result%7Bcolumn%3D%5C%22toy_value%5C%22%7D%22%2C+%22datasource%22%3A+%7B%22type%22%3A+%22prometheus%22%2C+%22uid%22%3A+%22prometheus%22%7D%7D%5D%2C+%22range%22%3A+%7B%22from%22%3A+%22now-15m%22%2C+%22to%22%3A+%22now%22%7D%7D%7D

aiops_python_result{column="toy_baseline"}
  http://localhost:3000/explore?schemaVersion=1&orgId=1&panes=%7B%22aiops%22%3A+%7B%22datasource%22%3A+%22prometheus%22%2C+%22queries%22%3A+%5B%7B%22refId%22%3A+%22A%22%2C+%22expr%22%3A+%22aiops_python_result%7Bcolumn%3D%5C%22toy_baseline%5C%22%7D%22%2C+%22datasource%22%3A+%7B%22type%22%3A+%22prometheus%22%2C+%22uid%22%3A+%22prometheus%22%7D%7D%5D%2C+%22range%22%3A+%7B%22from%22%3A+%22now-15m%22%2C+%22to%22%3A+%22now%22%7D%7D%7D

Lab 00 dashboard   http://localhost:3000/d/aiops-lab00


## 7. 自己動手做一個 panel

到目前為止你看的 panel 都是 `setup_grafana.py` 產生的。這一步要自己從零做一個，因為工作坊結束之後你面對的是自己公司的指標，那裡沒有腳本會幫你把 panel 生出來。

先在 Explore 裡把 query 調對，再放上 dashboard。這個順序值得養成習慣：Explore 是一個沒有後果的地方，query 打錯就重打，不會弄壞任何人正在看的 dashboard。上一格印的連結會直接把你送到填好 query 的 Explore，但下面的手動路徑要自己走一次，因為之後你面對的 metric 不會有人先幫你填好。

1. 左側導覽列 > **Explore**。
2. 左上角的 datasource 選 **prometheus**。
3. query 編輯器右上角有一個 **Builder / Code** 的切換，切到 **Code**。Builder 模式適合探索不熟的 metric，但它產生的 query 你看不到，而看不到的東西沒辦法帶去別的地方。
4. 輸入：

   ```promql
   aiops_python_result{column="toy_value"}
   ```

5. 右上角時間範圍設成 **Last 15 minutes**，按 **Run query**。圖出來了就表示 query 是對的。

   接下來把這段 query 變成 dashboard 上的一個 panel。

6. 左側導覽列 > **Dashboards** > **AIOps Workshop** > **Lab 00 - Pipeline check**。
7. 右上角 **Edit**，然後 **Add** > **Visualization**。
8. 新 panel 的 query 編輯器一樣切到 **Code**，把剛才那段 PromQL 貼進去。
9. 右側面板最上方把標題改成你自己認得的名字，例如 `my toy panel`。
10. 右側面板往下找 **Standard options** > **Unit**，搜尋 `short`，選 **Misc > short**。這一步不是裝飾。Grafana 預設不知道你的數字是什麼，軸上就只有裸數字；設了 unit 之後它才知道怎麼縮寫和怎麼標軸。
11. 再往下找 **Thresholds**，按 **+ Add threshold**，把值改成 `85`。
12. 回到 **Graph styles**，把 **Show thresholds** 設成 **As lines**。少了這一步，threshold 只會影響顏色，圖上不會有那條線。
13. 右上角 **Save dashboard**。

上面第 1 到第 3 步（Explore > datasource > **Code** 模式）是後面每一次寫 PromQL 的固定開場，Lab 01 和 Lab 02 都會叫你回到這個狀態，那時就不再重複寫一次。

做完之後把你的 panel 跟原本那個 **Toy signal published from the notebook** 並排看。兩者的 `toy_value` 應該一模一樣，差別在原本那個多畫了一條 `toy_baseline`，那是第二個 query。一個 panel 可以有好幾個 query，這是後面 Lab 01 畫 baseline 和上下界的做法。

## 8. 跟你自己這台機器比一比

dashboard 最下面兩個 panel，**Your own machine: receive rate** 和 **Your own machine: transmit rate**，資料來源完全不同。它們不經過 hop 1 和 hop 2，是 node_exporter 直接量你這台電腦的網卡，Prometheus 抓，Grafana 畫。三段而不是四段。

這兩個 panel 的存在有個具體理由。toy signal 是你自己造的，它會動是應該的。真實的網卡不受你控制，如果它也會動，就排除了「這條 pipeline 只對假資料有效」的可能。

panel 的 query 長這樣：

```promql
rate(node_network_receive_bytes_total{device!~"lo\d*|docker.*|veth.*|br-.*|utun.*|bridge.*|awdl.*|llw.*|gif.*|stf.*|anpi.*|ap[0-9].*"}[1m])
```

那一長串排除清單是必要的。一台 macOS 上 `node_network_receive_bytes_total` 有 19 條 series，絕大多數是 Apple 自己的虛擬介面，值永遠是 0，不濾掉的話圖上會有十幾條貼在零線上的線，把真正在動的那一條擠成看不見。實務上你會遇到同樣的問題，只是名字換成 `docker0` 和一堆 `veth`。

清單裡的 `lo\d*` 值得單獨看一眼，因為它是這個排除式最容易寫壞的地方。Prometheus 的 label 正規表示式是全字串錨定的，`=~` 和 `!~` 都要求整個值從頭到尾符合，不是找子字串。所以寫成 `lo` 只會命中名字剛好等於 `lo` 的介面，而 macOS 的 loopback 叫 `lo0`，整條漏過濾網。這不是紙上談兵。同一個瞬間量到 `en0` 76 kB/s、`lo0` 6.2 kB/s，loopback 在 panel 上是一條看得見的實線，機器閒著的時候它甚至可能是最高的那一條。`\d*` 把 `lo`、`lo0`、`lo1` 一起收掉。

濾完之後這台機器還留下 `en0` 到 `en5` 共 6 條，其中只有 `en0` 在動，`en1` 到 `en5` 量到的速率是 0，它們貼在零線上不影響判讀。排除清單要處理的是「十幾條假線蓋掉一條真線」，不是把 series 數壓到 1。

現在製造一點流量，然後看圖：

```bash
for i in $(seq 1 15); do curl -s -o /dev/null https://prometheus.io/docs/prometheus/latest/querying/basics/; done
```

同時開幾個瀏覽器分頁效果更明顯。看 receive rate 那個 panel，legend 上動的那條通常是 `en0`（Wi-Fi 或有線網卡）。

反應不是瞬間的，也不該是。`rate(...[1m])` 是在一分鐘的滑動視窗上算平均，所以流量結束之後，那個隆起還會花大約一分鐘才慢慢退回去。這個延遲是你選 `[1m]` 這個視窗長度換來的平滑。視窗愈長愈平滑，也愈遲鈍。Lab 01 挑 rolling window 長度時要做的是同一個取捨。

## 9. 在 Explore 裡把 PromQL 弄清楚

PromQL 的錯誤訊息只在 Prometheus 那一端看得到，包在 Python 裡查詢的話你會失去這個回饋，所以這一節全部在 Grafana Explore 裡做，一行 Python 都不寫。

回到第 7 節第 1 到第 3 步的狀態：Explore，datasource **prometheus**，**Code** 模式。以下每一段都貼進去按 Run，實際看回傳。

**instant vector**，最基本的形式。一個 metric 名字，回傳每一條 series 目前的值：

```promql
node_network_receive_bytes_total
```

在一台 macOS 上這會回 19 條 series，`en0` 那條的值大約是 13182692002。這個數字是自開機以來累積的位元組總數，只會往上，只有重開機或 counter 重置才會歸零。它本身沒有任何運維意義，沒有人會對「這張網卡這輩子收了 13 GB」設告警。

**label filter**，把 series 縮到你要的那一條。`=` 是精確比對，`=~` 是正規表示式：

```promql
node_network_receive_bytes_total{device="en0"}
node_network_receive_bytes_total{device=~"en.*"}
```

把累計 counter 換算成每秒變化量，靠的是 **rate()**：

```promql
rate(node_network_receive_bytes_total{device="en0"}[1m])
```

同一台機器上這回傳 85805，單位是 bytes/s。這才是可以拿來看趨勢、設門檻、做 baseline 的東西。`[1m]` 是 range selector，指定要往回看多久，`rate()` 拿這段區間的頭尾算斜率，同時會自動處理 counter reset。**對 counter 直接畫圖幾乎永遠是錯的，該畫的是它的 rate。**

**sum by**，沿著某些 label 聚合，其餘 label 丟掉：

```promql
sum by (instance) (rate(node_network_receive_bytes_total[1m]))
```

這會把一台機器上所有網卡的接收速率加起來，變成每台機器一條 series。`by` 保留列出的 label，`without` 則是反過來丟掉列出的。

**topk**，排序取前 N 名，找熱點時常用：

```promql
topk(3, rate(node_network_receive_bytes_total[5m]))
```

順帶一提，`up` 本身就是可以查的：

```promql
up
sum by (job) (up)
```

`up` 回傳每個 scrape target 一條 series，1 或 0。第 5 節的診斷就是建立在它身上。

### counter 和 delta 的差別，以及為什麼 Lab 01 依賴它

搞混這兩件事，Lab 01 的每一個 rolling window 都會算在錯的數列上。

`node_network_receive_bytes_total` 是**真正的 cumulative counter**。它單調遞增，所以必須用 `rate()` 才有意義。

課程的 CSV，`data/synthetic/synthetic_rrd_metrics.csv`，**不是**。裡面的 `INOCTETS`、`OUTOCTETS` 這些欄位存的是「前一個 polling interval 之內累積的量」，也就是 per-interval delta。它會隨著流量上上下下，不是單調的。

這兩者長得很像，都叫 octets，都是整數，都很大。搞混的代價是把訊號徹底破壞掉，而且不會有任何錯誤訊息。

`wk.load_telemetry()` 不猜。它用一個叫 `_monotonic_fraction` 的量測來判斷：計算同一個 port 上相鄰兩筆 `INOCTETS` 的差，統計其中有多少比例是非負的。真正的 counter 這個比例會非常接近 1.000，只有重置時才會出現負值。課程 CSV 實際量到的是 **0.503**，也就是差不多一半的相鄰差是負的，因為它跟著流量起伏。loader 據此判定為 delta 讀法，直接除以 cadence 300 秒得到 per-second rate，**完全不做 diff**。

loader 把這個判定寫成一份 contract，在 Lab 01 第一格執行 `wk.load_telemetry()` 之後印出來。

如果你硬是對這種 delta 欄位做 counter 讀法，`.diff()` 出來的樣本大約一半是負的，而任何合理的 counter 處理都會把負值 clip 成 0（`load_telemetry(semantics="counter")` 就是這樣做的，而且會在 contract 的 note 裡警告你）。剩下沒被 clip 掉的那一半，是兩個相鄰區間的差值，跟流量本身已經沒有關係。圖看起來還是會有東西，還是會有尖峰，只是它量的不再是你以為的東西。這種錯誤活得下來，因為它不會 crash。

順帶記住 cadence 是 300 秒不是 60 秒。一天 288 個樣本。任何「一列等於一分鐘」的心算都是錯的，Lab 01 算 rolling window 長度時第一個會踩到這裡。

## 10. 人的檢查點

先各自想過，再兩兩討論，最後全班對答案。沒有標準答案的部分比有標準答案的部分重要。

**第一題。** 你剛剛停掉了 exporter。斷的是哪一個 hop？如果沒有任何人告訴你發生了什麼事，你會從 dashboard 上的哪一個欄位、看到什麼樣的變化，而推論出這件事？

**第二題。** `Exporter up` 顯示 `0` 跟顯示 `No data`，分別代表什麼？哪一種比較難被發現？為什麼？在你自己公司的監控系統上，有沒有哪一個 panel 現在正處於第二種狀態而沒有人注意到？

**第三題。** 你在第 6 節改了 toy signal，panel 跟著動了。這證明了哪幾個 hop 是通的？而這個測試**沒有**覆蓋到什麼？（提示：Grafana 的 alert rule 走的是自己那條對 datasource 的查詢路徑，跟 dashboard panel 的查詢是兩件獨立的事。）

**第四題。** `node_network_receive_bytes_total` 需要 `rate()`，課程 CSV 的 `INOCTETS` 不需要。如果把兩者搞反，各自會在圖上看到什麼？哪一種錯誤比較容易被抓到？

## 卡住的時候

先跑 `wk.check_stack()`，再看第 5 節的症狀對照表，最後才回頭讀 Python。順序反過來的話，你會在一個沒有問題的 notebook 裡找一個不在那裡的 bug。

重啟 exporter 會讓 replay 從資料集的第一列重新開始，重啟 Prometheus 或 Grafana 則不會。無論重啟哪一個，新資料出現在 panel 上最多需要十幾秒。這不是故障。

下一站是 Lab 01，那裡開始問「跟什麼比」。